controllers/openstackfloatingippool_controller.go

```diff
@@ -0,0 +1,471 @@
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	scope, err := r.ScopeFactory.NewClientScopeFromFloatingIPPool(ctx, r.Client, pool, r.CaCertificates, log)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// This is done before deleting the pool, because we want to handle deleted IPs before we delete the pool
+	if err := r.reconcileIPAddresses(ctx, scope, pool); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if pool.ObjectMeta.DeletionTimestamp.IsZero() {
+		// Add finalizer if it does not exist
+		if controllerutil.AddFinalizer(pool, infrav1.OpenStackFloatingIPPoolFinalizer) {
+			return ctrl.Result{}, r.Client.Update(ctx, pool)
+		}
+	} else {
+		// Handle deletion
+		return ctrl.Result{}, r.reconcileDelete(ctx, scope, pool)
+	}
+
+	patchHelper, err := patch.NewHelper(pool, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		if err := patchHelper.Patch(ctx, pool); err != nil {
+			if reterr == nil {
+				reterr = fmt.Errorf("error patching OpenStackFloatingIPPool %s/%s: %w", pool.Namespace, pool.Name, err)
+			}
+		}
+	}()
+
+	if err := r.reconcileFloatingIPNetwork(scope, pool); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	claims := &ipamv1.IPAddressClaimList{}
+	if err := r.Client.List(context.Background(), claims, client.InNamespace(req.Namespace), client.MatchingFields{infrav1.OpenStackFloatingIPPoolNameIndex: pool.Name}); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	for _, claim := range claims.Items {
+		claim := claim
+		log := log.WithValues("claim", claim.Name)
+		if !claim.ObjectMeta.DeletionTimestamp.IsZero() {
+			continue
+		}
+
+		if claim.Status.AddressRef.Name == "" {
+			ipAddress := &ipamv1.IPAddress{}
+			err := r.Client.Get(ctx, client.ObjectKey{Name: claim.Name, Namespace: claim.Namespace}, ipAddress)
+			if client.IgnoreNotFound(err) != nil {
+				return ctrl.Result{}, err
+			}
+			if apierrors.IsNotFound(err) {
+				ip, err := r.getIP(scope, pool)
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+
+				ipAddress = &ipamv1.IPAddress{
+					ObjectMeta: ctrl.ObjectMeta{
+						Name:      claim.Name,
+						Namespace: claim.Namespace,
+						Finalizers: []string{
+							infrav1.DeleteFloatingIPFinalizer,
+						},
+						OwnerReferences: []metav1.OwnerReference{
+							{
+								APIVersion: claim.APIVersion,
+								Kind:       claim.Kind,
+								Name:       claim.Name,
+								UID:        claim.UID,
+							},
+						},
+					},
+					Spec: ipamv1.IPAddressSpec{
+						ClaimRef: corev1.LocalObjectReference{
+							Name: claim.Name,
+						},
+						PoolRef: corev1.TypedLocalObjectReference{
+							APIGroup: pointer.String(infrav1.GroupVersion.Group),
+							Kind:     pool.Kind,
+							Name:     pool.Name,
+						},
+						Address: ip,
+						Prefix:  32,
+					},
+				}
+
+				// Retry creating the IPAddress object
+				err = wait.ExponentialBackoff(backoff, func() (bool, error) {
```

Comment:
This should be cancellable:
```suggestion
				err = wait.ExponentialBackoffWithContext(ctx, backoff, func() (bool, error) {
```